File: Leetcode/3-Longest-Substring-Without-Repeating-Characters/LongestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var res = 0
    if (!s || s.length == 0) {
        return res;
    }
    
    var flag = new Array(256).fill(false)
    var start = 0
    var len = s.length
    var i
    for (i=0; i<len; i++) {
        var cur = s.charAt(i)
        var curIdx = cur.charCodeAt(0)
        if (!flag[curIdx]) {
            flag[curIdx] = true
            continue
        }
        res = Math.max(res, i-start)
        var j
        for (j=start; j<i; j++) {
            var curj = s.charAt(j)
            var curIdx = curj.charCodeAt(0)
            if (curj == cur) {
                start = j+1
                break;
            }
            flag[curIdx] = false
        }
    }
    res = Math.max(res, len - start)
    return res
};

File: Leetcode/3-Longest-Substring-Without-Repeating-Characters/LongestSubstringWithoutRepeatingCharacters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        if (s.length() == 0) {
            return res;
        }
        bool flag[256] = {false}; // Assuming ASCII
        int start = 0;
        int len = s.length();
        for (int i=0; i<len; i++) {
            char cur = s.at(i);
            if (!flag[cur]) {
                flag[cur] = true;
                continue;
            }
            res = max(res, i-start);
            for (int j=start; j<i; j++) {
                char c = s.at(j);
                if (c == cur) {
                    start = j+1;
                    break;
                }
                flag[c] = false;
            }
        }
        res = max(res, len - start);
        return res;
    }
};

